File: Algorithms/YandexTraining/1/5/3.py
N = int(input())
x = [0] * (N + 1)
y = [0] * (N + 1)

for i in range(1, N + 1):
    xi, yi = map(int, input().split())
    x[i] = xi
    y[i] = yi

up = [0] * (N + 1)   
down = [0] * (N + 1)

for i in range(2, N + 1):
    diff = y[i] - y[i - 1]
    up[i] = up[i - 1] + (diff if diff > 0 else 0)
    down[i] = down[i - 1] + ((-diff) if diff < 0 else 0)

M = int(input())
for _ in range(M):
    s, f = map(int, input().split())
    if s < f:
        print(up[f] - up[s])
    else:
        print(down[s] - down[f])

File: Algorithms/YandexTraining/1/5/3.cpp
#include <iostream>
#include <vector>

int main() {
    int N;
    std::cin >> N;

    std::vector<int> x(N + 1), y(N + 1);
    for (auto i = 1; i <= N; ++i) {
        std::cin >> x[i] >> y[i];
    }

    std::vector<int> up(N + 1, 0), down(N + 1, 0);
    for (auto i = 2; i <= N; ++i) {
        auto diff = y[i] - y[i - 1];
        up[i] = up[i - 1] + (diff > 0 ? diff : 0);
        down[i] = down[i - 1] + (diff < 0 ? -diff : 0);
    }

    int M;
    std::cin >> M;
    for (auto i = 0; i < M; ++i) {
        int s, f;
        std::cin >> s >> f;
        if (s < f) {
            std::cout << up[f] - up[s] << "\n";
        } else {
            std::cout << down[s] - down[f] << "\n";
        }
    }

    return 0;
}
